File: cpp/src/beam.cpp
#include <beam.h>

#include <string>
#include <limits>
#include <vector>

// JSON Keys
#define BACKAZIMUTH_KEY "BackAzimuth"
#define SLOWNESS_KEY "Slowness"
#define POWERRATIO_KEY "PowerRatio"
#define BACKAZIMUTHERROR_KEY "BackAzimuthError"
#define SLOWNESSERROR_KEY "SlownessError"
#define POWERRATIOERROR_KEY "PowerRatioError"

namespace detectionformats {
beam::beam() {
	backazimuth = std::numeric_limits<double>::quiet_NaN();
	backazimutherror = std::numeric_limits<double>::quiet_NaN();
	slowness = std::numeric_limits<double>::quiet_NaN();
	slownesserror = std::numeric_limits<double>::quiet_NaN();
	powerratio = std::numeric_limits<double>::quiet_NaN();
	powerratioerror = std::numeric_limits<double>::quiet_NaN();
}

beam::beam(double newbackazimuth, double newbackazimutherror,
			double newslowness, double newslownesserror, double newpowerratio,
			double newpowerratioerror) {
	backazimuth = newbackazimuth;
	backazimutherror = newbackazimutherror;
	slowness = newslowness;
	slownesserror = newslownesserror;
	powerratio = newpowerratio;
	powerratioerror = newpowerratioerror;
}

beam::beam(rapidjson::Value &json) {
	// required values
	// backazimuth
	if ((json.HasMember(BACKAZIMUTH_KEY) == true)
			&& (json[BACKAZIMUTH_KEY].IsNumber() == true)
			&& (json[BACKAZIMUTH_KEY].IsDouble() == true))
		backazimuth = json[BACKAZIMUTH_KEY].GetDouble();
	else
		backazimuth = std::numeric_limits<double>::quiet_NaN();

	// slowness
	if ((json.HasMember(SLOWNESS_KEY) == true)
			&& (json[SLOWNESS_KEY].IsNumber() == true)
			&& (json[SLOWNESS_KEY].IsDouble() == true))
		slowness = json[SLOWNESS_KEY].GetDouble();
	else
		slowness = std::numeric_limits<double>::quiet_NaN();

	// optional values
	// power ratio
	if ((json.HasMember(POWERRATIO_KEY) == true)
			&& (json[POWERRATIO_KEY].IsNumber() == true)
			&& (json[POWERRATIO_KEY].IsDouble() == true))
		powerratio = json[POWERRATIO_KEY].GetDouble();
	else
		powerratio = std::numeric_limits<double>::quiet_NaN();

	// backazimutherror
	if ((json.HasMember(BACKAZIMUTHERROR_KEY) == true)
			&& (json[BACKAZIMUTHERROR_KEY].IsNumber() == true)
			&& (json[BACKAZIMUTHERROR_KEY].IsDouble() == true))
		backazimutherror = json[BACKAZIMUTHERROR_KEY].GetDouble();
	else
		backazimutherror = std::numeric_limits<double>::quiet_NaN();

	// slownesserror
	if ((json.HasMember(SLOWNESSERROR_KEY) == true)
			&& (json[SLOWNESSERROR_KEY].IsNumber() == true)
			&& (json[SLOWNESSERROR_KEY].IsDouble() == true))
		slownesserror = json[SLOWNESSERROR_KEY].GetDouble();
	else
		slownesserror = std::numeric_limits<double>::quiet_NaN();

	// powerratioerror
	if ((json.HasMember(POWERRATIOERROR_KEY) == true)
			&& (json[POWERRATIOERROR_KEY].IsNumber() == true)
			&& (json[POWERRATIOERROR_KEY].IsDouble() == true))
		powerratioerror = json[POWERRATIOERROR_KEY].GetDouble();
	else
		powerratioerror = std::numeric_limits<double>::quiet_NaN();
}

beam::beam(const beam &newbeam) {
	backazimuth = newbeam.backazimuth;
	backazimutherror = newbeam.backazimutherror;
	slowness = newbeam.slowness;
	slownesserror = newbeam.slownesserror;
	powerratio = newbeam.powerratio;
	powerratioerror = newbeam.powerratioerror;
}

beam::~beam() {
}

rapidjson::Value & beam::tojson(
		rapidjson::Value &json,
		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
	json.SetObject();

	// required values
	// backazimuth
	if (std::isnan(backazimuth) != true)
		json.AddMember(BACKAZIMUTH_KEY, backazimuth, allocator);

	// slowness
	if (std::isnan(slowness) != true)
		json.AddMember(SLOWNESS_KEY, slowness, allocator);

	// optional values
	// powerratio
	if (std::isnan(powerratio) != true)
		json.AddMember(POWERRATIO_KEY, powerratio, allocator);

	// backazimutherror
	if (std::isnan(backazimutherror) != true)
		json.AddMember(BACKAZIMUTHERROR_KEY, backazimutherror, allocator);

	// slownesserror
	if (std::isnan(slownesserror) != true)
		json.AddMember(SLOWNESSERROR_KEY, slownesserror, allocator);

	// powerratioerror
	if (std::isnan(powerratioerror) != true)
		json.AddMember(POWERRATIOERROR_KEY, powerratioerror, allocator);

	return (json);
}

std::vector<std::string> beam::geterrors() {
	std::vector<std::string> errorlist;

	// check required data
	// backazimuth
	if (std::isnan(backazimuth) == true) {
		// backazimuth not found
		errorlist.push_back("No BackAzimuth in beam class.");
	}
	if (backazimuth < 0) {
		errorlist.push_back("Invalid BackAzimuth in beam class.");
	}

	// slowness
	if (std::isnan(slowness) == true) {
		// slowness not found
		errorlist.push_back("No Slowness in beam class.");
	}
	if (slowness < 0) {
		errorlist.push_back("Invalid Slowness in beam class.");
	}

	// optional data
	// powerratio
	if (std::isnan(powerratio) != true) {
		if (powerratio < 0) {
			errorlist.push_back("Invalid PowerRatio in beam class.");
		}
	}

	// backazimutherror
	if (std::isnan(backazimutherror) != true) {
		if (backazimutherror < 0) {
			errorlist.push_back("Invalid BackAzimuthError in beam class.");
		}
	}
	// slownesserror
	if (std::isnan(slownesserror) != true) {
		if (slownesserror < 0) {
			errorlist.push_back("Invalid SlownessError in beam class.");
		}
	}
	// powerratioerror
	if (std::isnan(powerratioerror) != true) {
		if (powerratioerror < 0) {
			errorlist.push_back("Invalid PowerRatioError in beam class.");
		}
	}

	// return the list of errors
	return (errorlist);
}

bool beam::isempty() {
	if (std::isnan(backazimuth) != true)
		return (false);
	if (std::isnan(slowness) != true)
		return (false);
	if (std::isnan(powerratio) != true)
		return (false);
	if (std::isnan(backazimutherror) != true)
		return (false);
	if (std::isnan(slownesserror) != true)
		return (false);
	if (std::isnan(powerratioerror) != true)
		return (false);

	return (true);
}
}  // namespace detectionformats

File: python/detectionformats/beam.py
#!/usr/bin/env python

#stdlib imports
import json

class Beam:
    """ Beam - a conversion class used to create, parse, and validate beam data 
        as part of detection data.
    """
    # json keys
    BACKAZIMUTH_KEY = "BackAzimuth"
    SLOWNESS_KEY = "Slowness"
    POWERRATIO_KEY = "PowerRatio"
    BACKAZIMUTHERROR_KEY = "BackAzimuthError"
    SLOWNESSERROR_KEY = "SlownessError"
    POWERRATIOERROR_KEY = "PowerRatioError"

    def __init__(self, newBackAzimuth=None, newSlowness=None, newPowerRatio=None,
        newBackAzimuthError=None, newSlownessError=None,
        newPowerRatioError=None):
        """Initialize the beam object. Constructs an empty object
           if all arguments are None

        Args:
            newBackAzimuth: a required Number containing the desired back
                azimuth measurement as a float
            newSlowness: a required Number containing the desired slowness
                measurement as a float
            newPowerRatio: an optional Number containing the desired power ratio
                measurement as a float
            newBackAzimuthError: an optional Number containing the desired back
                azimuth error measurement as a float
            newSlownessError: an optional Number containing the desired slowness
                error measurement as a float
            newPowerRatioError: an optional Number containing the desired power
                ratio error measurement as a float
        Returns:
            Nothing
        Raises:
            Nothing
        """
        # first required keys
        if newBackAzimuth is not None:
            self.backAzimuth = newBackAzimuth
        if newSlowness is not None:
            self.slowness = newSlowness

        # second optional keys
        if newPowerRatio is not None:
            self.powerRatio = newPowerRatio
        if newBackAzimuthError is not None:
            self.backAzimuthError = newBackAzimuthError
        if newSlownessError is not None:
            self.slownessError = newSlownessError
        if newPowerRatioError is not None:
            self.powerRatioError = newPowerRatioError

    def fromJSONString(self, jsonString):
        """Populates the object from a json formatted string

        Args:
            jsonString: a required String containing the json formatted text
        Returns:
            Nothing
        Raises:
            Nothing
        """
        jsonObject = json.loads(jsonString)
        self.fromDict(jsonObject)

    def fromDict(self, aDict):
        """Populates the object from a dictionary

        Args:
            aDict: a required dictionary
        Returns:
            Nothing
        Raises:
            Nothing
        """
        # first required keys
        try:
            self.backAzimuth = aDict[self.BACKAZIMUTH_KEY]
            self.slowness = aDict[self.SLOWNESS_KEY]
        except(ValueError, KeyError, TypeError) as e:
            print ("Dict format error, missing required keys: %s" % e)

        # second optional keys
        if self.POWERRATIO_KEY in aDict:
            self.powerRatio = aDict[self.POWERRATIO_KEY]
        if self.BACKAZIMUTHERROR_KEY in aDict:
            self.backAzimuthError = aDict[self.BACKAZIMUTHERROR_KEY]
        if self.SLOWNESSERROR_KEY in aDict:
            self.slownessError = aDict[self.SLOWNESSERROR_KEY]
        if self.POWERRATIOERROR_KEY in aDict:
            self.powerRatioError = aDict[self.POWERRATIOERROR_KEY]

    def toJSONString(self):
        """Converts the object to a json formatted string

        Args:
            None
        Returns:
            The JSON formatted message as a String
        Raises:
            Nothing
        """
        jsonObject = self.toDict()

        return json.dumps(jsonObject, ensure_ascii=False)

    def toDict(self):
        """Converts the object to a dictionary

        Args:
            None
        Returns:
            The dictionary
        Raises:
            Nothing
        """
        aDict = {}

        # first required keys
        try:
            aDict[self.BACKAZIMUTH_KEY] = self.backAzimuth
            aDict[self.SLOWNESS_KEY] = self.slowness
        except(NameError, AttributeError) as e:
            print ("Missing required data error: %s" % e)

        # second optional keys
        if hasattr(self, 'powerRatio'):
            aDict[self.POWERRATIO_KEY] = self.powerRatio

        if hasattr(self, 'backAzimuthError'):
            aDict[self.BACKAZIMUTHERROR_KEY] = self.backAzimuthError

        if hasattr(self, 'slownessError'):
            aDict[self.SLOWNESSERROR_KEY] = self.slownessError

        if hasattr(self, 'powerRatioError'):
            aDict[self.POWERRATIOERROR_KEY] = self.powerRatioError

        return aDict

    def isValid(self):
        """Checks to see if the object is valid

        Args:
            None
        Returns:
            True if the object is valid, False otherwise
        Raises:
            Nothing
        """
        errorList = self.getErrors()

        return not errorList

    def getErrors(self):
        """Gets a list of object validation errors

        Args:
            None
        Returns:
            A List of Strings containing the validation error messages
        Raises:
            Nothing
        """
        errorList = []

        # first required keys
        try:
            if self.backAzimuth < 0:
                errorList.append('Invalid BackAzimuth in Beam Class.')
        except(NameError, AttributeError):
            errorList.append('No BackAzimuth in Beam Class.')

        try:
            if self.slowness < 0:
                errorList.append('Invalid Slowness in Beam Class.')
        except(NameError, AttributeError):
            errorList.append('No Slowness in Beam Class.')

        # second optional keys
        if hasattr(self, 'powerRatio'):
            if self.powerRatio < 0:
                errorList.append('Invalid PowerRatio in Beam Class.')

        if hasattr(self, 'backAzimuthError'):
            if self.backAzimuthError < 0:
                errorList.append('Invalid BackAzimuthError in Beam Class.')

        if hasattr(self, 'slownessError'):
            if self.slownessError < 0:
                errorList.append('Invalid SlownessError in Beam Class.')

        if hasattr(self, 'powerRatioError'):
            if self.powerRatioError < 0:
                errorList.append('Invalid PowerRatioError in Beam Class.')

        return errorList

    def isEmpty(self):
        """Checks to see if the object is empty

        Args:
            None
        Returns:
            True if the object is empty, False otherwise
        Raises:
            Nothing
        """
        if hasattr(self, 'backAzimuth'):
            return False

        if hasattr(self, 'slowness'):
            return False

        if hasattr(self, 'powerRatio'):
            return False

        if hasattr(self, 'backAzimuthError'):
            return False

        if hasattr(self, 'slownessError'):
            return False

        if hasattr(self, 'powerRatioError'):
            return False           

        return True
